README.md
# scroll:bit

 [![Build Status](https://travis-ci.org/pimoroni/pxt-scrollbit.svg?branch=master)](https://travis-ci.org/pimoroni/pxt-scrollbit)

17x7 pixels of single-colour, brightness-controlled, message scrolling goodness!

This package adds support for the Pimoroni scroll:bit to makecode.microbit.org.

A scroll:bit is required to use this package, grab yours here: https://shop.pimoroni.com/products/scroll-bit

To use this package, go to https://makecode.microbit.org/, click "Advanced" then "Add Package" and search for scroll:bit. 

## JavaScript Reference

scroll:bit has 119 pixels arranged in 17 columns and 7 rows, each pixel supports 255 levels of brightness although these are gamma-corrected to 183 possible values.

### Get the size of your scroll:bit

The blocks `cols` and `rows` return the width and height of scroll:bit in pixels respectively.

Use these in your loops and flow control to draw amazing things!

### Light a single pixel

Light a single pixel on scroll:bit. Note, you must call `show` to display your changes.

* `col` is the column, from 0-16
* `row` is the row, from 0-6
* `brightness` is the brightness, from 0-255

```
scrollbit.setPixel(col: number, row:number, brightness: number)
```

For example:

```
scrollbit.setPixel(5, 5, 255)
```

### Display your changes

When you've finished setting pixels and drawing text, you must call `show` to display your changes.

```
scrollbit.show()
```

### Clear the display

To clear the display, you can call `clear`, you must also call `show` if you want to display your changes.

```
scrollbit.clear()
scrollbit.show()
```

### Display a text string

To show a string of text on scroll:bit you should use `drawText`:

* `col` is the column, from 0-16
* `row` is the row, from 0-6
* `text` is the text you want to show
* `brightness` is the brightness, from 0-255

```
scrollbit.drawText(col: number, row:number, text: string, brightness: number)
```

For example:

```
scrollbit.drawText(0, 1, "Hello World", 128)
```

### Scroll a text string across the display

Since a long string of text wont fit on scroll:bit, you have to scroll it across the display.

You can do this manually with `drawText` and `measureText` but we've included a convinient function for you:

* `text` is the text you want to show
* `brightness` is the brightness, from 0-255
* `delay` (optional) is the time, in milliseconds, to delay each scroll step - a bigger number equals slower scrolling

```
scrollbit.scrollText(text: string, brightness: number, delay: number=50)
```

For example:

```
scrollbit.scrollText("The quick brown fox jumped over the lazy dog!", 128)
```

Scrolling text is always vertically centered, starts off the right edge of the display, and scrolls across until it's gone.

### Draw a single character

If you just want to draw a single lett, you can use `drawChar`:

* `col` is the column, from 0-16
* `row` is the row, from 0-6
* `char` is the text you want to show
* `brightness` is the brightness, from 0-255

```
scrollbit.drawChar(col: number, row: number, char: string, brightness: number)
```

For example:

```
scrollbit.drawChar(1, 1, 'H', 128)
scrollbit.drawChar(7, 1, 'a', 128)
scrollbit.drawChar(12,1, 'i', 128)
```

### Measure a text string

It can be useful to know how long a string of text might be, in pixels, on scroll:bit. Use `measureText` to find out:

```
scrollbit.measureText(text: string)
```

For example:

```
let width: number = scrollbit.measureText("Hello World")
```

This will return a number of pixels corresponding to the length of the text as it's displayed on scroll:bit (using the built-in 5x5 micro:bit font).

### Icons & Arrows

You can use icons and arrows in your text, just place their name in curly brackets like so: `"Hello {Heart} World"` or: `"Boo! Went the {Ghost}"` or: `"{Heart}{SmallHeart}{Heart} Happy Birthday! {Heart}{SmallHeart}{Heart}"`

Here's a list of icons you can use:

* Heart
* SmallHeart
* Yes
* No
* Happy
* Sad
* Confused
* Angry
* Asleep
* Surprised
* Silly
* Fabulous
* Meh
* TShirt
* Rollerskate
* Duck
* House
* Tortoise
* Butterfly
* StickFigure
* Ghost
* Sword
* Giraffe
* Skull
* Umbrella
* Snake
* Rabbit
* Cow
* QuarterNote
* EigthNote
* Pitchfork
* Target
* Triangle
* LeftTriangle
* Chessboard
* Diamond
* SmallDiamond
* Square
* SmallSquare
* Scissors
* North
* NorthEast
* East
* SouthEast
* South
* SouthWest
* West
* NorthWest

## License

MIT License

Copyright (c) 2018 Pimoroni Ltd.

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.

## Supported targets

* for PXT/microbit

```package
scrollbit=github:pimoroni/pxt-scrollbit
```


CHANGELOG.md
# 0.0.7

* Compatibility fix for new micro:bit V2

# 0.0.6

* @kshoji - Fix bug with missing bounds check
* @septolum - Add support for 180 degree rotation

# 0.0.5

* Add support for `Image` and `Big Image` blocks
* Move gamma correction into C bindings, freeing up resources
* Bumped pxt-microbit to 1.2.18

# 0.0.4

* Fix mismatch between release version and pxt.json

# 0.0.3

* Forward-compatibility fixes for new micro:bit editor

# 0.0.2

* Repository and Travis fixes

# 0.0.1

* Initial release


test.ts
// tests go here; this will not be compiled when this package is used as a library

/*
basic.forever(() => {
    scrollbit.scrollText("{Heart}")
    scrollbit.scrollText("{SmallHeart}")
    scrollbit.scrollText("{Yes}")
    scrollbit.scrollText("{No}")
    scrollbit.scrollText("{Happy}")
    scrollbit.scrollText("{Sad}")
    scrollbit.scrollText("{Confused}")
    scrollbit.scrollText("{Angry}")
    scrollbit.scrollText("{Asleep}")
    scrollbit.scrollText("{Surprised}")
    scrollbit.scrollText("{Silly}")
    scrollbit.scrollText("{Fabulous}")
    scrollbit.scrollText("{Meh}")
    scrollbit.scrollText("{TShirt}")
    scrollbit.scrollText("{Rollerskate}")
    scrollbit.scrollText("{Duck}")
    scrollbit.scrollText("{House}")
    scrollbit.scrollText("{Tortoise}")
    scrollbit.scrollText("{Butterfly}")
    scrollbit.scrollText("{StickFigure}")
    scrollbit.scrollText("{Ghost}")
    scrollbit.scrollText("{Sword}")
    scrollbit.scrollText("{Giraffe}")
    scrollbit.scrollText("{Skull}")
    scrollbit.scrollText("{Umbrella}")
    scrollbit.scrollText("{Snake}")
    scrollbit.scrollText("{Rabbit}")
    scrollbit.scrollText("{Cow}")
    scrollbit.scrollText("{QuarterNote}")
    scrollbit.scrollText("{EightNote}")
    scrollbit.scrollText("{Pitchfork}")
    scrollbit.scrollText("{Target}")
    scrollbit.scrollText("{Triangle}")
    scrollbit.scrollText("{LeftTriangle}")
    scrollbit.scrollText("{Chessboard}")
    scrollbit.scrollText("{Diamond}")
    scrollbit.scrollText("{SmallDiamond}")
    scrollbit.scrollText("{Square}")
    scrollbit.scrollText("{SmallSquare}")
    scrollbit.scrollText("{Scissors}")
    scrollbit.scrollText("{North}")
    scrollbit.scrollText("{NorthEast}")
    scrollbit.scrollText("{East}")
    scrollbit.scrollText("{SouthEast}")
    scrollbit.scrollText("{South}")
    scrollbit.scrollText("{SouthWest}")
    scrollbit.scrollText("{West}")
    scrollbit.scrollText("{NorthW}")
})
*/

/*

basic.showIcon(IconNames.Heart)

let brightness: number = 128

for (let col: number = 0; col < scrollbit.cols(); col++){
    for (let row: number = 0; row < scrollbit.rows(); row++){
        let offset: number = (row * scrollbit.cols()) + col
        let br: number = (brightness + 25) - ((offset * brightness) / 119)
        br = Math.max(0, br)
        if (offset > 117) br = 0
        scrollbit.setPixel(col, row, br)
    }
}

scrollbit.show()
*/

/*
// Test Jig Code

basic.forever(function(){

    for (let col: number = 0; col < scrollbit.cols(); col++){
        for (let row: number = 0; row < scrollbit.rows(); row++){
            let offset: number = (col * scrollbit.cols()) + row
            scrollbit.setPixel(col, row, 64 * (offset % 2))
        }
    }
    scrollbit.show()
    control.waitMicros(1000000)
    for (let col: number = 0; col < scrollbit.cols(); col++){
        for (let row: number = 0; row < scrollbit.rows(); row++){
            let offset: number = (col * scrollbit.cols()) + row
            scrollbit.setPixel(col, row, 64 * ((offset + 1) % 2))
        }
    }
    scrollbit.show()
    control.waitMicros(1000000)
    for (let col: number = 0; col < scrollbit.cols(); col++){
        for (let row: number = 0; row < scrollbit.rows(); row++){
            let offset: number = (col * scrollbit.cols()) + row
            scrollbit.setPixel(col, row, 64)
        }
    }
    scrollbit.show()
    control.waitMicros(1000000)
    scrollbit.clear()    
    scrollbit.show()
    control.waitMicros(1000000)

})
*/

/*

scrollbit.scrollText("Yarrrrrrrrrrrrrrrrrr")
scrollbit.scrollText("The quick, brown fox jumped over the lazy dog! How slow does this get when displaying a really long sentence? I wonder. Let's see how far we can push this thing.", 128, 0)
scrollbit.scrollText("Hello World! How are you today?", 128)
scrollbit.scrollText("Super duper scrollalicious!", 128)
scrollbit.setPixel(0,0,128)
scrollbit.show()
*/

font.cpp
#include "pxt.h"
using namespace pxt;

#ifndef PXT_CREATE_BUFFER
#define PXT_CREATE_BUFFER(data, len) ManagedBuffer(data, len).leakData()
#endif


namespace scrollbit {
    //%
    int getFontDataByte(int index) {
        if(index < 0 || index >= 475){
            return 0;
        }
        #if MICROBIT_CODAL
            auto font = codal::BitmapFont::getSystemFont();
        #else
            auto font = MicroBitFont::getSystemFont();
        #endif
        return (char)*(font.characters + index);
    }
    //%
    int getCharWidth(int charCode) {
        if(charCode < MICROBIT_FONT_ASCII_START || charCode > MICROBIT_FONT_ASCII_END){
            return 5;
        }
        #if MICROBIT_CODAL
            auto font = codal::BitmapFont::getSystemFont();
        #else
            auto font = MicroBitFont::getSystemFont();
        #endif
        int offset = (charCode - MICROBIT_FONT_ASCII_START) * 5;
        uint8_t width = (uint8_t)*(font.characters + offset)
        | (uint8_t)*(font.characters + offset + 1)
        | (uint8_t)*(font.characters + offset + 2)
        | (uint8_t)*(font.characters + offset + 3)
        | (uint8_t)*(font.characters + offset + 4);

        if (width & 1) { return 5; }
        if (width & 2) { return 4; }
        if (width & 4) { return 3; }
        return 2;
    }
    //%
    Buffer getFontData(int charCode) {
        if(charCode < MICROBIT_FONT_ASCII_START || charCode > MICROBIT_FONT_ASCII_END){
            return PXT_CREATE_BUFFER(NULL, 5);
        }
        #if MICROBIT_CODAL
            auto font = codal::BitmapFont::getSystemFont();
        #else
            auto font = MicroBitFont::getSystemFont();
        #endif
        int offset = (charCode - MICROBIT_FONT_ASCII_START) * 5;

        return PXT_CREATE_BUFFER((uint8_t *)(font.characters + offset), 5);
    }
}

gamma.cpp
#include "pxt.h"
using namespace pxt;

namespace scrollbit {
    const unsigned char gamma[256] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2,
        2, 2, 2, 3, 3, 3, 3, 3, 4, 4, 4, 4, 5, 5, 5, 5,
        6, 6, 6, 7, 7, 7, 8, 8, 8, 9, 9, 9, 10, 10, 11, 11,
        11, 12, 12, 13, 13, 13, 14, 14, 15, 15, 16, 16, 17, 17, 18, 18,
        19, 19, 20, 21, 21, 22, 22, 23, 23, 24, 25, 25, 26, 27, 27, 28,
        29, 29, 30, 31, 31, 32, 33, 34, 34, 35, 36, 37, 37, 38, 39, 40,
        40, 41, 42, 43, 44, 45, 46, 46, 47, 48, 49, 50, 51, 52, 53, 54,
        55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70,
        71, 72, 73, 74, 76, 77, 78, 79, 80, 81, 83, 84, 85, 86, 88, 89,
        90, 91, 93, 94, 95, 96, 98, 99, 100, 102, 103, 104, 106, 107, 109, 110,
        111, 113, 114, 116, 117, 119, 120, 121, 123, 124, 126, 128, 129, 131, 132, 134,
        135, 137, 138, 140, 142, 143, 145, 146, 148, 150, 151, 153, 155, 157, 158, 160,
        162, 163, 165, 167, 169, 170, 172, 174, 176, 178, 179, 181, 183, 185, 187, 189,
        191, 193, 194, 196, 198, 200, 202, 204, 206, 208, 210, 212, 214, 216, 218, 220,
        222, 224, 227, 229, 231, 233, 235, 237, 239, 241, 244, 246, 248, 250, 252, 255};

    //%
    int correctGamma(int brightness)
    {
        return gamma[brightness & 0xff];
    }
}

scrollbit.ts
//% weight=100 color=#000000 icon="\uf06e" block="Scroll:Bit"
namespace scrollbit {
    const I2C_ADDR: number = 0x74
    const REG_MODE: number = 0x00
    const REG_FRAME: number = 0x01
    const REG_AUDIOSYNC: number = 0x06
    const REG_SHUTDOWN: number = 0x0a

    const REG_COLOR: number = 0x24
    const REG_BLINK: number = 0x12
    const REG_ENABLE: number = 0x00

    const REG_BANK: number = 0xfd
    const BANK_CONFIG: number = 0x0b

    const COLS: number = 17
    const ROWS: number = 7

    let UPSIDE_DOWN: boolean = false

    const ARROWOFFSET: number = 40

    const ICONS: string[] = [
        "Heart",
        "SmallH",
        "Yes",
        "No",
        "Happy",
        "Sad",
        "Confus",
        "Angry",
        "Asleep",
        "Surpri",
        "Silly",
        "Fabulo",
        "Meh",
        "TShirt",
        "Roller",
        "Duck",
        "House",
        "Tortoi",
        "Butter",
        "StickF",
        "Ghost",
        "Sword",
        "Giraff",
        "Skull",
        "Umbrel",
        "Snake",
        "Rabbit",
        "Cow",
        "Quarte",
        "EigthN",
        "Pitchf",
        "Target",
        "Triang",
        "LeftTr",
        "Chessb",
        "Diamon",
        "SmallD",
        "Square",
        "SmallS",
        "Scisso",
    
        "North",
        "NorthE",
        "East",
        "SouthE",
        "South",
        "SouthW",
        "West",
        "NorthW"
    ]

    let buf: Buffer = pins.createBuffer(144)
    let frame: number = 0

    /**
     * Scroll text across scroll:bit
     * @param text - text to scroll
     * @param brightness - brightness to set (0-255)
     * @param delay - additional delay in milliseconds (0-100)
     */
    //% blockId=scrollbit_scroll_text 
    //% block="scroll %text| at brightness %brightness| with delay (ms) %delay"
    //% brightness.min=0 brightness.max=255 brightness.defl=128
    //% delay.min=0 delay.max=100 delay.defl=50
    export function scrollText(text: string, brightness: number = 128, delay: number = 50) {
        text = tokenize(text)
        let len: number = measureText(text)
        for (let col: number = 0; col < len + COLS; col++){
            clear()
            _drawText(text, -col + COLS, 1, brightness)
            show()
            if (delay > 0) {
                control.waitMicros(delay * 1000)
            }    
        }
    }

    function tokenize(text: string): string {
        let result: string = ""
        let icon: string = ""

        for (let x = 0; x < text.length; x++){
            let char: string = text.charAt(x)
            if (char == "}" && icon.length > 0) {
                let index: number = ICONS.indexOf(icon.substr(1,6))
                icon += char

                if (index > -1) {
                    icon = String.fromCharCode(DAL.MICROBIT_FONT_ASCII_END + 1 + index)
                }

                result += icon
                icon = ""
                continue
            }
            if (char == "{" || icon.length > 0) {
                icon += char
                continue
            }
            result += char
        }

        return result
    }

    /**
     * Display an icon on scroll:bit
     * @param icon - icon to display
     * @param col - column to set (0-16)
     * @param row - row to set (0-6)
     * @param brightness - brightness to set (0-255)
     */
    //% blockId=scrollbit_set_icon
    //% block="display icon %icon| at col %col| row %row| with brightness %brightness"
    //% icon.fieldEditor="gridpicker"
    //% icon.fieldOptions.width="400" icon.fieldOptions.columns="5"
    //% icon.fieldOptions.itemColour="black" icon.fieldOptions.tooltips="true"
    //% col.min=0 col.max=16
    //% row.min=0 row.max=6
    //% brightness.min=0 brightness.max=255 brightness.defl=128
    export function setIcon(icon: IconNames, col: number, row: number, brightness: number = 128): void {
        let image: Image = images.iconImage(icon)
        setImage(image, col, row, brightness)
    }

    /**
     * Display an arrow on scroll:bit
     * @param arrow - arrow to display
     * @param col - column to set (0-16)
     * @param row - row to set (0-6)
     * @param brightness - brightness to set (0-255)
     */
    //% blockId=scrollbit_set_arrow
    //% block="display arrow %arrow| at col %col| row %row| with brightness %brightness"
    //% col.min=0 col.max=16
    //% row.min=0 row.max=6
    //% brightness.min=0 brightness.max=255 brightness.defl=128
    export function setArrow(arrow: ArrowNames, col: number, row: number, brightness: number = 128): void {
        let image: Image = images.arrowImage(arrow)
        setImage(image, col, row, brightness)
    }

    /**
     * Display an image on scroll:bit
     * @param image - image to display
     * @param col - column to set (0-16)
     * @param row - row to set (0-6)
     * @param brightness - brightness to set (0-255)
     */
    //% blockId=scrollbit_set_image
    //% block="display image %image| at col %col| row %row| with brightness %brightness"
    //% col.min=0 col.max=16
    //% row.min=0 row.max=6
    //% brightness.min=0 brightness.max=255 brightness.defl=128
    export function setImage(image: Image, col: number, row: number, brightness: number = 128): void {
        for (let c_row = 0; c_row < 5; c_row++) {
            for (let c_col = 0; c_col < image.width(); c_col++) {
                if (image.pixelBrightness(c_col, c_row)) {
                    setPixel(col + c_col, row + c_row, brightness)
                }
            }
        }
    }

    /**
     * Set a single pixel on scroll:bit
     * @param col - column to set (0-16)
     * @param row - row to set (0-6)
     * @param brightness - brightness to set (0-255)
     */
    //% blockId=scrollbit_set_pixel
    //% block="set pixel| at col %col| row %row| to %brightness"
    //% col.min=0 col.max=16
    //% row.min=0 row.max=6
    //% brightness.min=0 brightness.max=255 brightness.defl=128
    export function setPixel(col: number, row: number, brightness: number = 128): void {
        if(col < 0 || row < 0 || col >= COLS || row >= ROWS){return}
        if (UPSIDE_DOWN) {col = (COLS - 1) - col; row = (ROWS - 1) - row}
        buf[pixelAddr(col, row)] = Math.clamp(0, 255, brightness)
    }

    /**
     * Set the orientation of the scroll:bit
     * @param upside_down - is the scroll:bit upside down? (true/false)
     */
    //% blockId=scrollbit_upside_down
    //% block="turn the scroll:bit| upside down? %upside_down"
    //% upside_down.defl=false
    export function setUpsideDown(upside_down: boolean): void {
        UPSIDE_DOWN = upside_down
    }

    /**
     * Update scroll:bit,
     * update the display with your pretty pixels
     */
    //% blockId=scrollbit_show
    //% block="display your changes"
    export function show(): void {
        let corrected_buf: Buffer = pins.createBuffer(144)

        for (let x = 0; x < buf.length; x++) {
            corrected_buf[x] = correctGamma(buf[x])
        }

        smbus.writeByte(I2C_ADDR, REG_BANK, frame)
        smbus.writeBuffer(I2C_ADDR, REG_COLOR, corrected_buf)
        smbus.writeByte(I2C_ADDR, REG_BANK, BANK_CONFIG)
        smbus.writeByte(I2C_ADDR, REG_FRAME, frame)

        frame = frame == 0 ? 1 : 0
    }

    /**
     * Clear scroll:bit,
     * sets all pixels to 0, but does not update display
     */
    //% blockId=scrollbit_clear
    //% block="clear scroll:bit"
    //% icon=""
    export function clear(): void {
        buf.fill(0)
    }

    /**
     * Set all pixels on the scroll:bit
     * @param brightness - brightness to set (0-255)
     */
    //% blockId=scrollbit_set_all
    //% block="set all pixels| to %brightness"
    //% brightness.min=0 brightness.max=255 brightness.defl=128
    export function setAll(brightness: number = 128): void {
        buf.fill(Math.clamp(0, 255, brightness))
    }

    /**
     * Return the width (number of cols) of scroll:bit
     */
    //% blockId=scrollbit_cols
    //% block="number of columns"
    //% color=#444444
    //% icon=""
    export function cols(): number {
        return COLS
    }

    /**
     * Return the height (number of rows) of scroll:bit
     */
    //% blockId=scrollbit_rows
    //% block="number of rows"
    //% color=#444444
    //% icon=""
    export function rows(): number {
        return ROWS
    }

    /**
     * Get a single pixel on scroll:bit,
     * returns a brightness value (0-255)
     * @param col - column to get (0-16)
     * @param row - row to get (0-6)
     */
    //% blockId=scrollbit_get_pixel icon="\uf0eb"
    //% block="get pixel| at col %col| row %row"
    //% advanced color=#554444
    export function getPixel(col: number, row: number): number {
        return buf[pixelAddr(col, row)]
    }

    /**
     * Measure text, returns a length in pixels
     * @param text - text to measure
     */
    //% blockId=scrollbit_measure_text
    //% block="get length of %text in pixels"
    //% advanced color=#554444
    export function measureText(text: string): number {
        let len: number = 0
        for (let x: number = 0; x < text.length; x++){
            len += charWidth(text.charAt(x)) + 1
        }
        return len
    }

    /**
     * Draw a single alphanumeric character.
     * @param col - column position (0-16)
     * @param row - row position (0-6)
     * @param char - character to display
     * @param brightness - brightness (0-255)
     */
    export function drawChar(char: string, col: number, row: number, brightness: number = 128): void {
        if (char.charCodeAt(0) > DAL.MICROBIT_FONT_ASCII_END + ARROWOFFSET) {
            setArrow(char.charCodeAt(0) - DAL.MICROBIT_FONT_ASCII_END - ARROWOFFSET - 1, col, row, brightness)
            return;
        }
        if (char.charCodeAt(0) > DAL.MICROBIT_FONT_ASCII_END) {
            setIcon(char.charCodeAt(0) - DAL.MICROBIT_FONT_ASCII_END - 1, col, row, brightness)
            return;
        }
        let data: Buffer = getChar(char)
        for (let c_row = 0; c_row < 5; c_row++) {
            for (let c_col = 0; c_col < 5; c_col++) {
                if ((data[c_row] & (1 << (4 - c_col))) > 0) {
                    setPixel(col + c_col, row + c_row, brightness)
                }
            }
        }
    }

    /**
     * Draw text on scroll:bit
     * @param col - column to set (0-16)
     * @param row - row to set (0-6)
     * @param text - text to show
     * @param brightness - brightness to set (0-255)
     */
    //% blockId=scrollbit_draw_text
    //% block="draw %text| at col %col| row %row| with brightness %brightness"
    //% advanced color=#220000
    //% col.min=0 col.max=16
    //% row.min=0 row.max=6
    //% brightness.min=0 brightness.max=255 brightness.defl=128
    export function drawText(text: string, col: number, row: number, brightness: number = 128): void {
        text = tokenize(text)
        _drawText(text, col, row, brightness)
    }
    export function _drawText(text: string, col: number, row: number, brightness: number = 128): void {
        let offset_col: number = 0
        for (let x: number = 0; x < text.length; x++){
            let width:  number = charWidth(text.charAt(x))
            if (col + offset_col >= COLS) {
                return
            }
            if (col + offset_col + width < 0) {
                offset_col += width + 1
                continue
            }
            drawChar(text.charAt(x), col + offset_col, row, brightness)
            offset_col += width + 1
        }
    }

    /**
     * TODO: Get some GUI/UX magic for this
     * Set all the pixels on scroll:bit using
     * a text string. Brightness = charCode.
     * eg ' ' = 32, 'z' = 122
     */
    //% blockId=scrollbit_show_leds
    //% block="display LEDs from text %text"
    //% advanced color=#220000
    export function showLeds(leds: string): void {
        leds = leds.substr(0, COLS * ROWS)
        for (let x = 0; x < leds.length; x++) {
            let brightness = leds.charCodeAt(x);
            let col = x % COLS
            let row = Math.idiv((x - col), COLS)
            setPixel(col, row, brightness)
        }
        show()
    }

    function pixelAddr(col: number, row: number): number {
        row = 7 - (row + 1)

        if (col > 8) {
            col = col - 8
            row = 6 - (row + 8)
        } else {
            col = 8 - col
        }

        return (col * 16) + row
    }

    //% shim=scrollbit::correctGamma
    function correctGamma(brightness: number): number {
        return 0
    }


    //% shim=scrollbit::getFontDataByte
    function getFontDataByte(index: number): number {
        return 0
    }

    //% shim=scrollbit::getFontData
    function getFontData(index: number): Buffer {
        return pins.createBuffer(5)
    }

    //% shim=scrollbit::getCharWidth
    function getCharWidth(char: number): number {
        return 5
    }

    function getChar(character: string): Buffer {
        return getFontData(character.charCodeAt(0))
    }

    function charWidth(character: string): number {
        let charcode: number = character.charCodeAt(0)
        if (charcode > DAL.MICROBIT_FONT_ASCII_END) {
            return 5
        }
        return getCharWidth(charcode)
    }

    /**
     * Setup scroll:bit,
     * must be called before using the display
     */
    export function setup(): void {
        clear()
        smbus.writeByte(I2C_ADDR, REG_BANK, BANK_CONFIG)

        control.waitMicros(1000)
        smbus.writeByte(I2C_ADDR, REG_SHUTDOWN, 0)
        control.waitMicros(1000)
        smbus.writeByte(I2C_ADDR, REG_SHUTDOWN, 1)
        control.waitMicros(1000)

        smbus.writeByte(I2C_ADDR, REG_MODE, 0)
        smbus.writeByte(I2C_ADDR, REG_AUDIOSYNC, 0)

        let enable = pins.createBuffer(17)
        enable.fill(255)
        smbus.writeByte(I2C_ADDR, REG_BANK, 0)
        smbus.writeBuffer(I2C_ADDR, 0, enable)
        smbus.writeByte(I2C_ADDR, REG_BANK, 1)
        smbus.writeBuffer(I2C_ADDR, 0, enable)
    }
}

namespace smbus {
    export function writeByte(addr: number, register: number, value: number): void {
        let temp = pins.createBuffer(2);
        temp[0] = register;
        temp[1] = value;
        pins.i2cWriteBuffer(addr, temp, false);
    }
    export function writeBuffer(addr: number, register: number, value: Buffer): void {
        let temp = pins.createBuffer(value.length + 1);
        temp[0] = register;
        for (let x = 0; x < value.length; x++) {
            temp[x + 1] = value[x];
        }
        pins.i2cWriteBuffer(addr, temp, false);
    }
    export function readBuffer(addr: number, register: number, len: number): Buffer {
        let temp = pins.createBuffer(1);
        temp[0] = register;
        pins.i2cWriteBuffer(addr, temp, false);
        return pins.i2cReadBuffer(addr, len, false);
    }
    function readNumber(addr: number, register: number, fmt: NumberFormat = NumberFormat.UInt8LE): number {
        let temp = pins.createBuffer(1);
        temp[0] = register;
        pins.i2cWriteBuffer(addr, temp, false);
        return pins.i2cReadNumber(addr, fmt, false);
    }
}

scrollbit.setup();
